libgalois/include/katana/GraphTopology.h

```diff
@@ -0,0 +1,1175 @@
+
+  void invalidate() noexcept { is_valid_ = false; }
+
+  bool has_edges_sorted_by(const EdgeSortKind& kind) const noexcept {
+    if (kind == EdgeSortKind::kAny) {
+      return true;
+    }
+    return edge_sort_state_ == kind;
+  }
+
+  const PropertyGraph& property_graph() const noexcept { return *prop_graph_; }
+
+  PropertyIndex edge_property_index(const Edge& eid) const noexcept {
+    KATANA_LOG_DEBUG_ASSERT(eid < num_edges());
+    return edge_prop_indices_[eid];
+  }
+
+  void sortEdges(const EdgeSortKind& edge_sort_todo) noexcept {
+    switch (edge_sort_todo) {
+    case EdgeSortKind::kAny:
+      return;
+    case EdgeSortKind::kSortedByDestID:
+      SortEdgesByDestID();
+      return;
+    case EdgeSortKind::kSortedByEdgeType:
+      SortEdgesByTypeThenDest();
+      return;
+    case EdgeSortKind::kSortedByNodeType:
+      KATANA_LOG_FATAL("Not implemented yet");
+      return;
+    default:
+      KATANA_LOG_FATAL("switch-case fell through");
+      return;
+    }
+  }
+
+  static std::unique_ptr<EdgeShuffleTopology> MakeTransposeCopy(
+      const PropertyGraph* pg);
+  static std::unique_ptr<EdgeShuffleTopology> MakeOriginalCopy(
+      const PropertyGraph* pg);
+
+  static std::unique_ptr<EdgeShuffleTopology> Make(
+      const PropertyGraph* pg, const TransposeKind& tpose_todo,
+      const EdgeSortKind& edge_sort_todo) noexcept {
+    std::unique_ptr<EdgeShuffleTopology> ret;
+
+    if (tpose_todo == TransposeKind::kYes) {
+      ret = MakeTransposeCopy(pg);
+      KATANA_LOG_DEBUG_ASSERT(ret->has_transpose_state(TransposeKind::kYes));
+    } else {
+      ret = MakeOriginalCopy(pg);
+      KATANA_LOG_DEBUG_ASSERT(ret->has_transpose_state(TransposeKind::kNo));
+    }
+
+    ret->sortEdges(edge_sort_todo);
+    return ret;
+  }
+
+  void SortEdgesByDestID() noexcept;
+
+  void SortEdgesByTypeThenDest() noexcept;
+
+  void SortEdgesByDestType(const PropIndexVec& node_prop_indices) noexcept;
+
+protected:
+  EdgeShuffleTopology(
+      const PropertyGraph* pg, const TransposeKind& tpose_todo,
+      const EdgeSortKind& edge_sort_todo, AdjIndexVec&& adj_indices,
+      EdgeDestVec&& dests, PropIndexVec&& edge_prop_indices) noexcept
+      : Base(std::move(adj_indices), std::move(dests)),
+        prop_graph_(pg),
+        is_valid_(true),
+        tpose_state_(tpose_todo),
+        edge_sort_state_(edge_sort_todo),
+        edge_prop_indices_(std::move(edge_prop_indices)) {
+    KATANA_LOG_DEBUG_ASSERT(edge_prop_indices_.size() == num_edges());
+  }
+
+private:
+  const PropertyGraph* prop_graph_;
+  bool is_valid_ = true;
+  TransposeKind tpose_state_ = TransposeKind::kNo;
+  EdgeSortKind edge_sort_state_ = EdgeSortKind::kAny;
+
+  // TODO(amber): In the future, we may need to keep a copy of edge_type_ids in
+  // addition to edge_prop_indices_. Today, we assume that we can use
+  // PropertyGraph.edge_type_set_id(edge_prop_indices_[edge_id]) to obtain
+  // edge_type_id. This may not be true when we group properties
```

Comment:
edge_type_id will also be grouped in that case. If we group by type-id, then we don't need to maintain edge_type_id